Mathlib/Data/Bitvec/Defs.lean

```diff
@@ -1,339 +1,177 @@
-  Sle y x
-#align bitvec.sge Bitvec.Sge
-
-end Comparison
-
-/-! ### Conversion to `nat` and `int` -/
-
-
-section Conversion
-
-variable {α : Type}
-
-/-- Create a bitvector from a `nat` -/
-protected def ofNat : ∀ n : ℕ, Nat → Bitvec n
-  | 0, _ => nil
-  | succ n, x => Bitvec.ofNat n (x / 2)++ₜdecide (x % 2 = 1) ::ᵥ nil
-#align bitvec.of_nat Bitvec.ofNat
-
-/-- Create a bitvector from an `Int`. The ring homomorphism from Int to bitvectors. -/
-protected def ofInt : ∀ n : ℕ, Int → Bitvec n
-  | n, Int.ofNat m => Bitvec.ofNat n m
-  | n, Int.negSucc m => (Bitvec.ofNat n m).not
-
-/-- `add_lsb r b` is `r + r + 1` if `b` is `true` and `r + r` otherwise. -/
-def addLsb (r : ℕ) (b : Bool) :=
-  r + r + cond b 1 0
-#align bitvec.add_lsb Bitvec.addLsb
-
-/-- Given a `List` of `Bool`s, return the `nat` they represent as a list of binary digits. -/
-def bitsToNat (v : List Bool) : Nat :=
-  v.foldl addLsb 0
-#align bitvec.bits_to_nat Bitvec.bitsToNat
-
-/-- Return the natural number encoded by the input bitvector -/
-protected def toNat {n : Nat} (v : Bitvec n) : Nat :=
-  bitsToNat (toList v)
-#align bitvec.to_nat Bitvec.toNat
-
-instance (n : ℕ) : Preorder (Bitvec n) :=
-  Preorder.lift Bitvec.toNat
-
-/-- convert `fin` to `Bitvec` -/
-def ofFin {n : ℕ} (i : Fin <| 2 ^ n) : Bitvec n :=
-  Bitvec.ofNat _ i.val
-#align bitvec.of_fin Bitvec.ofFin
-
-/-- convert `Bitvec` to `fin` -/
-def toFin {n : ℕ} (i : Bitvec n) : Fin (2 ^ n) :=
-  i.toNat
-#align bitvec.to_fin Bitvec.toFin
-
-
-/-- Return the integer encoded by the input bitvector -/
-protected def toInt : ∀ {n : Nat}, Bitvec n → Int
-  | 0, _ => 0
-  | succ _, v =>
-    cond (head v) (Int.negSucc <| Bitvec.toNat <| Bitvec.not <| tail v)
-      (Int.ofNat <| Bitvec.toNat <| tail v)
-#align bitvec.to_int Bitvec.toInt
-
-end Conversion
-
-/-! ### Miscellaneous instances -/
-
-
-private def repr {n : Nat} : Bitvec n → String
-  | ⟨bs, _⟩ => "0b" ++ (bs.map fun b : Bool => if b then '1' else '0').asString
-
-instance (n : Nat) : Repr (Bitvec n) where
-  reprPrec (b : Bitvec n) _ := Std.Format.text (repr b)
-
-end Bitvec
-
-instance {n} {x y : Bitvec n} : Decidable (Bitvec.Ult x y) :=
-  decEq _ _
+/-!
+## Arithmetic
+We inherit `Fin` implementations.
+-/
 
-instance {n} {x y : Bitvec n} : Decidable (Bitvec.Ugt x y) :=
-  decEq _ _
+/-- Less than for two bitvectors. -/
+protected def lt (x y : BitVec w) : Bool :=
+  x.val < y.val
 
-instance {n} : HShiftLeft (Bitvec n) Nat (Bitvec n) := ⟨Bitvec.shl⟩
+/-- Less than or equal to for two bitvectors. -/
+protected def le (x y : BitVec w) : Bool :=
+  x.val ≤ y.val
 
-instance {n} : HShiftRight (Bitvec n) Nat (Bitvec n) := ⟨Bitvec.ushr⟩
+instance : Ring (BitVec w) := inferInstanceAs (Ring (Fin _))
+instance : Mod (BitVec w) := inferInstanceAs (Mod (Fin _))
+instance : Div (BitVec w) := inferInstanceAs (Div (Fin _))
+instance : LT (BitVec w)  := ⟨fun x y => BitVec.lt x y⟩
+instance : LE (BitVec w)  := ⟨fun x y => BitVec.le x y⟩
```

Comment:
Why not inherit the `LinearOrder` instance from `Fin` here?

Perhaps a better question: is there anything you *don't* want to inherit from `Fin`? If not, making `BitVec` an `abbrev` would save you a lot of work.